src/stan/variational/advi.hpp

```diff
@@ -0,0 +1,665 @@
+        Eigen::VectorXd mu_grad =
+          Eigen::VectorXd::Zero(model_.num_params_r());
+        Eigen::VectorXd sigma_tilde_grad =
+          Eigen::VectorXd::Zero(model_.num_params_r());
+
+        // ADAgrad parameters
+        double tau = 1.0;
+        Eigen::VectorXd mu_s =
+          Eigen::VectorXd::Zero(model_.num_params_r());
+        Eigen::VectorXd sigma_tilde_s =
+          Eigen::VectorXd::Zero(model_.num_params_r());
+
+        // RMSprop window_size
+        double window_size = 100.0;
+        double post_factor = 1.0 / window_size;
+        double pre_factor  = 1.0 - post_factor;
+
+        // Initialize ELBO and convergence tracking variables
+        double elbo(0.0);
+        double elbo_prev = std::numeric_limits<double>::min();
+        double delta_elbo = std::numeric_limits<double>::max();
+        double delta_elbo_ave = std::numeric_limits<double>::max();
+        double delta_elbo_med = std::numeric_limits<double>::max();
+
+        // Heuristic to estimate how far to look back in rolling window
+        int cb_size = (int)
+                std::max(0.1*max_iterations/static_cast<double>(refresh_), 1.0);
+        boost::circular_buffer<double> cb(cb_size);
+
+        // Print stuff
+        *print_stream_ << "  iter"
+                       << "       ELBO"
+                       << "   delta_ELBO_mean"
+                       << "   delta_ELBO_med"
+                       << "   notes "
+                       << std::endl;
+
+        // Timing variables
+        clock_t start = clock();
+        clock_t end;
+        double delta_t;
+
+        // Main loop
+        std::vector<double> print_vector;
+        bool do_more_iterations = true;
+        int iter_counter = 0;
+        while (do_more_iterations) {
+          // Compute gradient using Monte Carlo integration
+          calc_combined_grad(musigmatilde, mu_grad, sigma_tilde_grad);
+
+          // Accumulate S vector for ADAgrad
+          mu_s.array()           += mu_grad.array().square();
+          sigma_tilde_s.array()  += sigma_tilde_grad.array().square();
+
+          // RMSprop moving average weighting
+          mu_s.array() = pre_factor * mu_s.array()
+                       + post_factor * mu_grad.array().square();
+          sigma_tilde_s.array() = pre_factor * sigma_tilde_s.array()
+                                + post_factor
+                                  * sigma_tilde_grad.array().square();
+
+          // Take ADAgrad or rmsprop step
+          musigmatilde.set_mu(
+            musigmatilde.mu().array() +
+            eta_stepsize_ * mu_grad.array()
+            / (tau + mu_s.array().sqrt()));
+          musigmatilde.set_sigma_tilde(
+            musigmatilde.sigma_tilde().array()  +
+            eta_stepsize_ * sigma_tilde_grad.array()
+            / (tau + sigma_tilde_s.array().sqrt()));
+
+          // Check for convergence every "refresh_"th iteration
+          if (iter_counter % refresh_ == 0) {
+            elbo_prev = elbo;
```

Comment:
@akucukelbir Given heuristics of convergence, it might make sense to put the following 5-10 lines into a function that gets called internally to the fullrank and meanfield run_robbins_munro_adagrad. That way if we decide to change the heuristics, we can do it once easily.